apollo-router-core/src/layers/deduplication.rs

```diff
@@ -1,130 +1,339 @@
     async fn dedup(
-        service: S,
-        wait_map: WaitMap,
-        request: SubgraphRequest,
-    ) -> Result<SubgraphResponse, BoxError> {
+        mut service: S,
+        wait_map: Arc<Mutex<HashMap<Key, Sender<Result<S::Response, String>>>>>,
+        key: Key,
+        request_fn: fn(&Key) -> Request,
+    ) -> Result<S::Response, BoxError> {
         loop {
             let mut locked_wait_map = wait_map.lock().await;
-            match locked_wait_map.get_mut(&request.http_request) {
+            let waiter = locked_wait_map.get_mut(&key);
+            match waiter {
                 Some(waiter) => {
                     // Register interest in key
                     let mut receiver = waiter.subscribe();
                     drop(locked_wait_map);
 
                     match receiver.recv().await {
-                        Ok(value) => {
-                            return value
-                                .map(|response| SubgraphResponse {
-                                    response: response.response,
-                                    context: request.context,
-                                })
-                                .map_err(|e| e.into())
-                        }
+                        Ok(value) => return value.map_err(Into::into),
                         // there was an issue with the broadcast channel, retry fetching
                         Err(_) => continue,
                     }
                 }
                 None => {
                     let (tx, _rx) = broadcast::channel(1);
-                    locked_wait_map.insert(request.http_request.clone(), tx.clone());
+                    locked_wait_map.insert(key.clone(), tx.clone());
                     drop(locked_wait_map);
 
-                    let context = request.context.clone();
-                    let http_request = request.http_request.clone();
-                    let res = service.ready_oneshot().await?.call(request).await;
+                    let downstream_request = (request_fn)(&key);
+                    let res = service.call(downstream_request).await;
 
                     {
                         let mut locked_wait_map = wait_map.lock().await;
-                        locked_wait_map.remove(&http_request);
+                        locked_wait_map.remove(&key);
                     }
 
                     // Let our waiters know
-                    let broadcast_value = res
-                        .as_ref()
-                        .map(|response| response.clone())
-                        .map_err(|e| e.to_string());
+                    // We map the error into a string because we can't guarantee the error type is clone.
+                    let value = res.map_err(|e| Into::<BoxError>::into(e).to_string());
 
                     // Our use case is very specific, so we are sure that
                     // we won't get any errors here.
-                    tokio::task::spawn_blocking(move || {
```

Comment:
@garypen Why was broadcasting the result done from a thread previously?

Reply:
The send() method is blocking and amounts to cloning and sending the value for each subscriber, so it can block the executor for some time. It's safer to isolate that in spawn_blocking

Reply:
Added a comment around this.
All broadcast channel does when sending is to update a mutex and notify waiting threads to wake. Introducing this extra thread is the thing that makes the difference in performance between the current implementation and the previous.

Reply:
awesome! so my assumptions about broadcast were wrong :sweat_smile: